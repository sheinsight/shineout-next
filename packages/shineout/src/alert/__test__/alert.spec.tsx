import { render, cleanup, fireEvent, waitFor } from '@testing-library/react';
import '@testing-library/jest-dom';
import Alert from '..';
import mountTest from '../../tests/mountTest';
import { classLengthTest } from '../../tests/structureTest';
import {
  attributesTest,
  baseTest,
  childrenTest,
  classTest,
  delay,
  displayTest,
  snapshotTest,
  styleTest,
  textContentTest,
} from '../../tests/utils';
import AlertBase from '../__example__/01-base';
import AlertType from '../__example__/02-type';
import AlertCloseable from '../__example__/03-closeable';
import AlertTitle from '../__example__/04-title';

const SO_PREFIX = 'alert';
const alertClassName = `.${SO_PREFIX}-alert-0-2-1`;
const alertInfoClassName = `${SO_PREFIX}-info-0-2-7`;
const alertSuccessClassName = `${SO_PREFIX}-success-0-2-8`;
const alertWarningClassName = `${SO_PREFIX}-warning-0-2-9`;
const alertDangerClassName = `${SO_PREFIX}-danger-0-2-10`;
const alertIconClassName = `.${SO_PREFIX}-icon-0-2-12`;
const alertCloseClassName = `.${SO_PREFIX}-close-0-2-4`;
const alertPendingClassName = `${SO_PREFIX}-pending-0-2-6`;
const alertNoBorderedClassName = `${SO_PREFIX}-noBordered-0-2-11`;
const alertWidthTitleClassName = `${SO_PREFIX}-widthTitle-0-2-2`;
const alertTitleClassName = `.${SO_PREFIX}-title-0-2-3`;

afterEach(cleanup);
mountTest(<Alert />);
describe('Alert[Base]', () => {
  displayTest(Alert, 'ShineoutAlert');
  baseTest(Alert, alertClassName);
  childrenTest(Alert, alertClassName);
  snapshotTest(<AlertBase />);
  snapshotTest(<AlertType />, 'about type');
  snapshotTest(<AlertCloseable />, 'about closeable');
  snapshotTest(<AlertTitle />, 'about title');
  test('should render default', () => {
    const { container } = render(<Alert>Demo</Alert>);
    const alert = container.querySelector(alertClassName)!;
    classTest(alert, alertWarningClassName);
    expect(alert.firstChild).toHaveAttribute('data-soui-layout', 'row');
    const content = alert.firstChild?.firstChild;
    expect(content).toHaveAttribute('data-soui-children', 'true');
    expect(content).toHaveAttribute('data-soui-layout', 'cell');
    textContentTest(content as Element, 'Demo');
  });
  test('should render when set icon', () => {
    const { container } = render(<Alert icon>Demo</Alert>);
    const alert = container.querySelector(alertClassName)!;
    const icon = alert.querySelector(alertIconClassName)!;
    expect(icon).toBeInTheDocument();
    attributesTest(icon.parentElement!, 'data-soui-icon', 'true');
    attributesTest(icon.parentElement!, 'data-soui-layout', 'cell');
    classLengthTest(icon, 'svg', 1);
  });
  test('should render when set iconSize', () => {
    const { container } = render(
      <Alert icon iconSize={20}>
        Demo
      </Alert>,
    );
    styleTest(container.querySelector(alertIconClassName)!, 'width: 20px;');
  });
  test('should render when icon is element', () => {
    const iconDemo = <i>i</i>;
    const { container } = render(<Alert icon={iconDemo}>Demo</Alert>);
    const alert = container.querySelector(alertClassName)!;
    const icon = alert.querySelector(alertIconClassName)!;
    classLengthTest(icon, 'i', 1);
  });
  test('should render when set bordered is fasle', () => {
    const { container } = render(<Alert bordered={false}>Demo</Alert>);
    const alert = container.querySelector(alertClassName)!;
    classTest(alert, alertNoBorderedClassName);
  });
});
describe('Alert[Type]', () => {
  test('should render when set different type', () => {
    const classes = [
      alertInfoClassName,
      alertSuccessClassName,
      alertWarningClassName,
      alertDangerClassName,
    ];
    const { container } = render(<AlertType />);
    container.querySelectorAll(alertClassName).forEach((item, index) => {
      classTest(item, classes[index]);
    });
  });
});
describe('Alert[Close]', () => {
  test('should render when set closable', () => {
    const { container } = render(<Alert closable>Demo</Alert>);
    const alert = container.querySelector(alertClassName)!;
    const close = alert.querySelector(alertCloseClassName)!;
    expect(close).toBeInTheDocument();
    attributesTest(close.parentElement!, 'data-soui-close', 'true');
    classLengthTest(close, 'svg', 1);
  });
  // TODO: closeFn不支持boolean
  test('should render when set close', async () => {
    const closeFn = jest.fn();
    const { container } = render(<Alert onClose={closeFn}>Demo</Alert>);
    const alert = container.querySelector(alertClassName)!;
    const close = alert.querySelector(alertCloseClassName)!;
    expect(close).toBeInTheDocument();
    fireEvent.click(close);
    classTest(alert, alertPendingClassName);
    await waitFor(async () => {
      await delay(400);
      expect(alert).not.toBeInTheDocument();
      expect(closeFn.mock.calls.length).toBe(1);
    });
  });
  test('should render when set hideClose', () => {
    const closeFn = jest.fn();
    const { container } = render(
      <Alert hideClose onClose={closeFn}>
        Demo
      </Alert>,
    );
    const alert = container.querySelector(alertClassName)!;
    classLengthTest(alert, alertCloseClassName, 0);
  });
  test('should render when set close and closable at the same time', () => {
    const closeFn = jest.fn();
    const { container } = render(
      <Alert closable={false} onClose={closeFn}>
        Demo
      </Alert>,
    );
    const alert = container.querySelector(alertClassName)!;
    const close = alert.querySelector(alertCloseClassName)!;
    expect(close).not.toBeInTheDocument();
  });
  // TODO: closeItem
});
describe('Alert[Title]', () => {
  test('should render when set title', () => {
    const titleDemo = <div className='demo'>Hello</div>;
    const { container } = render(<Alert title={titleDemo}>Demo</Alert>);
    const alert = container.querySelector(alertClassName)!;
    classTest(alert, alertWidthTitleClassName);
    const title = alert.querySelector(alertTitleClassName)!;
    expect(title).toBeInTheDocument();
    classLengthTest(title, '.demo', 1);
    textContentTest(title, 'Hello');
  });
});
