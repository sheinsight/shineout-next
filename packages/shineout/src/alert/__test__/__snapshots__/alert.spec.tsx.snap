// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Alert[Base] should render correctly  1`] = `
<div
  class="alert-alert-0-2-1 alert-info-0-2-7"
>
  <div
    data-soui-layout="row"
  >
    <div
      data-soui-icon="true"
      data-soui-layout="cell"
    >
      <div
        class="alert-icon-0-2-12"
      >
        <svg
          aria-hidden="true"
          fill="currentColor"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M12 23C5.92487 23 1 18.0751 1 12C1 5.92487 5.92487 1 12 1C18.0751 1 23 5.92487 23 12C23 18.0751 18.0751 23 12 23ZM12 8C12.5523 8 13 7.55228 13 7C13 6.44772 12.5523 6 12 6C11.4477 6 11 6.44772 11 7C11 7.55228 11.4477 8 12 8ZM12 19C12.5523 19 13 18.5523 13 18V10C13 9.44772 12.5523 9 12 9C11.4477 9 11 9.44772 11 10V18C11 18.5523 11.4477 19 12 19Z"
          />
        </svg>
      </div>
    </div>
    <div
      data-soui-children="true"
      data-soui-layout="cell"
    >
      This is a line of important text for alerting purposes
    </div>
  </div>
</div>
`;

exports[`Alert[Base] should render correctly about closeable 1`] = `
<div
  class="alert-alert-0-2-1 alert-info-0-2-7"
>
  <div
    data-soui-layout="row"
  >
    <div
      data-soui-icon="true"
      data-soui-layout="cell"
    >
      <div
        class="alert-icon-0-2-12"
      >
        <svg
          aria-hidden="true"
          fill="currentColor"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M12 23C5.92487 23 1 18.0751 1 12C1 5.92487 5.92487 1 12 1C18.0751 1 23 5.92487 23 12C23 18.0751 18.0751 23 12 23ZM12 8C12.5523 8 13 7.55228 13 7C13 6.44772 12.5523 6 12 6C11.4477 6 11 6.44772 11 7C11 7.55228 11.4477 8 12 8ZM12 19C12.5523 19 13 18.5523 13 18V10C13 9.44772 12.5523 9 12 9C11.4477 9 11 9.44772 11 10V18C11 18.5523 11.4477 19 12 19Z"
          />
        </svg>
      </div>
    </div>
    <div
      data-soui-children="true"
      data-soui-layout="cell"
    >
      This is informative text.
    </div>
    <div
      data-soui-close="true"
      data-soui-layout="cell"
    >
      <div
        class="alert-close-0-2-4"
      >
        <svg
          aria-hidden="true"
          fill="currentColor"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M4.6129 3.2097L4.70711 3.29289L12 10.585L19.2929 3.29289C19.6834 2.90237 20.3166 2.90237 20.7071 3.29289C21.0676 3.65338 21.0953 4.22061 20.7903 4.6129L20.7071 4.70711L13.415 12L20.7071 19.2929C21.0976 19.6834 21.0976 20.3166 20.7071 20.7071C20.3466 21.0676 19.7794 21.0953 19.3871 20.7903L19.2929 20.7071L12 13.415L4.70711 20.7071C4.31658 21.0976 3.68342 21.0976 3.29289 20.7071C2.93241 20.3466 2.90468 19.7794 3.2097 19.3871L3.29289 19.2929L10.585 12L3.29289 4.70711C2.90237 4.31658 2.90237 3.68342 3.29289 3.29289C3.65338 2.93241 4.22061 2.90468 4.6129 3.2097Z"
          />
        </svg>
      </div>
    </div>
  </div>
</div>
`;

exports[`Alert[Base] should render correctly about title 1`] = `
<div
  class="alert-alert-0-2-1 alert-info-0-2-7 alert-widthTitle-0-2-2"
>
  <div
    data-soui-layout="row"
  >
    <div
      data-soui-icon="true"
      data-soui-layout="cell"
    >
      <div
        class="alert-icon-0-2-12"
      >
        <svg
          aria-hidden="true"
          fill="currentColor"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M12 23C5.92487 23 1 18.0751 1 12C1 5.92487 5.92487 1 12 1C18.0751 1 23 5.92487 23 12C23 18.0751 18.0751 23 12 23ZM12 8C12.5523 8 13 7.55228 13 7C13 6.44772 12.5523 6 12 6C11.4477 6 11 6.44772 11 7C11 7.55228 11.4477 8 12 8ZM12 19C12.5523 19 13 18.5523 13 18V10C13 9.44772 12.5523 9 12 9C11.4477 9 11 9.44772 11 10V18C11 18.5523 11.4477 19 12 19Z"
          />
        </svg>
      </div>
    </div>
    <div
      data-soui-layout="cell"
    >
      <div
        class="alert-title-0-2-3"
      >
        Heading
      </div>
    </div>
  </div>
  <div
    data-soui-layout="row"
  >
    <div
      data-soui-layout="cell"
    />
    <div
      data-soui-children="true"
      data-soui-layout="cell"
    >
      This is a line of important text for alerting purposes
    </div>
  </div>
</div>
`;

exports[`Alert[Base] should render correctly about type 1`] = `
<div
  style="display: flex; flex-wrap: wrap; gap: 32px;"
>
  <div
    class="alert-alert-0-2-1 alert-info-0-2-7"
  >
    <div
      data-soui-layout="row"
    >
      <div
        data-soui-icon="true"
        data-soui-layout="cell"
      >
        <div
          class="alert-icon-0-2-12"
        >
          <svg
            aria-hidden="true"
            fill="currentColor"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M12 23C5.92487 23 1 18.0751 1 12C1 5.92487 5.92487 1 12 1C18.0751 1 23 5.92487 23 12C23 18.0751 18.0751 23 12 23ZM12 8C12.5523 8 13 7.55228 13 7C13 6.44772 12.5523 6 12 6C11.4477 6 11 6.44772 11 7C11 7.55228 11.4477 8 12 8ZM12 19C12.5523 19 13 18.5523 13 18V10C13 9.44772 12.5523 9 12 9C11.4477 9 11 9.44772 11 10V18C11 18.5523 11.4477 19 12 19Z"
            />
          </svg>
        </div>
      </div>
      <div
        data-soui-children="true"
        data-soui-layout="cell"
      >
        This is informative text.
      </div>
    </div>
  </div>
  <div
    class="alert-alert-0-2-1 alert-success-0-2-8"
  >
    <div
      data-soui-layout="row"
    >
      <div
        data-soui-icon="true"
        data-soui-layout="cell"
      >
        <div
          class="alert-icon-0-2-12"
        >
          <svg
            aria-hidden="true"
            fill="currentColor"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M12 23C5.92487 23 1 18.0751 1 12C1 5.92487 5.92487 1 12 1C18.0751 1 23 5.92487 23 12C23 18.0751 18.0751 23 12 23ZM6.34315 13.8284L8.46305 15.9483C9.24511 16.7304 10.5122 16.7304 11.2943 15.9483L17.6569 9.58579C18.0472 9.19546 18.0472 8.5619 17.6569 8.17157C17.2665 7.78125 16.633 7.78125 16.2426 8.17157L10.5872 13.827C10.1962 14.218 9.56119 14.218 9.17016 13.827L7.75736 12.4142C7.36704 12.0239 6.73347 12.0239 6.34315 12.4142C5.95282 12.8045 5.95282 13.4381 6.34315 13.8284Z"
            />
          </svg>
        </div>
      </div>
      <div
        data-soui-children="true"
        data-soui-layout="cell"
      >
        This is success text.
      </div>
    </div>
  </div>
  <div
    class="alert-alert-0-2-1 alert-warning-0-2-9"
  >
    <div
      data-soui-layout="row"
    >
      <div
        data-soui-icon="true"
        data-soui-layout="cell"
      >
        <div
          class="alert-icon-0-2-12"
        >
          <svg
            aria-hidden="true"
            fill="currentColor"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M12 23C5.92487 23 1 18.0751 1 12C1 5.92487 5.92487 1 12 1C18.0751 1 23 5.92487 23 12C23 18.0751 18.0751 23 12 23ZM12 8C12.5523 8 13 7.55228 13 7C13 6.44772 12.5523 6 12 6C11.4477 6 11 6.44772 11 7C11 7.55228 11.4477 8 12 8ZM12 19C12.5523 19 13 18.5523 13 18V10C13 9.44772 12.5523 9 12 9C11.4477 9 11 9.44772 11 10V18C11 18.5523 11.4477 19 12 19Z"
            />
          </svg>
        </div>
      </div>
      <div
        data-soui-children="true"
        data-soui-layout="cell"
      >
        This is warning text.
      </div>
    </div>
  </div>
  <div
    class="alert-alert-0-2-1 alert-danger-0-2-10"
  >
    <div
      data-soui-layout="row"
    >
      <div
        data-soui-icon="true"
        data-soui-layout="cell"
      >
        <div
          class="alert-icon-0-2-12"
        >
          <svg
            aria-hidden="true"
            fill="currentColor"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M12 1C18.0751 1 23 5.92487 23 12C23 18.0751 18.0751 23 12 23C5.92487 23 1 18.0751 1 12C1 5.92487 5.92487 1 12 1ZM9.7845 8.38125C9.39221 8.07623 8.82498 8.10396 8.46449 8.46444C8.07397 8.85497 8.07397 9.48813 8.46449 9.87866L10.5846 11.9995L8.46449 14.1213L8.3813 14.2155C8.07628 14.6078 8.10401 15.175 8.46449 15.5355C8.85502 15.926 9.48818 15.926 9.87871 15.5355L11.9996 13.4136L14.1213 15.5355L14.2156 15.6187C14.6078 15.9237 15.1751 15.896 15.5356 15.5355C15.9261 15.145 15.9261 14.5118 15.5356 14.1213L13.4136 11.9995L15.5356 9.87866L15.6187 9.78445C15.9238 9.39216 15.896 8.82493 15.5356 8.46444C15.145 8.07392 14.5119 8.07392 14.1213 8.46444L11.9996 10.5845L9.87871 8.46444L9.7845 8.38125Z"
            />
          </svg>
        </div>
      </div>
      <div
        data-soui-children="true"
        data-soui-layout="cell"
      >
        This is danger text.
      </div>
    </div>
  </div>
</div>
`;
